File: platform/polyglot/jira_api/py/src/jira_api/models/project.py
"""Pydantic models for JIRA project entities."""

from __future__ import annotations

from datetime import datetime
from typing import Optional

from pydantic import BaseModel, Field


class ProjectVersion(BaseModel):
    """JIRA Project Version."""

    id: str
    name: str
    description: str | None = None
    archived: bool = False
    released: bool = False
    start_date: datetime | None = None
    release_date: datetime | None = None
    overdue: bool | None = None
    user_start_date: str | None = None
    user_release_date: str | None = None
    project_id: int = Field(..., alias="projectId")

    model_config = {"populate_by_name": True}


class ProjectVersionCreate(BaseModel):
    """Model for creating a new project version."""

    name: str
    description: str | None = None
    project_id: int
    archived: bool = False
    released: bool = False
    start_date: str | None = None
    release_date: str | None = None


class IssueType(BaseModel):
    """JIRA Issue Type."""

    id: str
    name: str
    description: str = ""
    icon_url: str | None = None
    subtask: bool = False


class ProjectLead(BaseModel):
    """JIRA Project Lead."""

    account_id: str | None = Field(None, alias="accountId")
    display_name: str | None = Field(None, alias="displayName")
    active: bool | None = None
    avatar_urls: dict[str, str] | None = Field(None, alias="avatarUrls")

    model_config = {"populate_by_name": True}


class Project(BaseModel):
    """JIRA Project."""

    id: str
    key: str
    name: str
    description: str | None = None
    lead: ProjectLead | None = None
    project_type_key: str | None = Field(None, alias="projectTypeKey")
    avatar_urls: dict[str, str] | None = Field(None, alias="avatarUrls")
    url: str | None = None
    issue_types: list[IssueType] | None = Field(None, alias="issueTypes")
    versions: list[ProjectVersion] | None = None

    model_config = {"populate_by_name": True}


class ProjectDetails(BaseModel):
    """Extended project details."""

    project: Project
    versions: list[ProjectVersion] = Field(default_factory=list)
    issue_types: list[IssueType] = Field(default_factory=list)

File: platform/polyglot/jira_api/mjs/src/models/project.mjs
/**
 * @module models/project
 * @description Zod schemas and types for JIRA project entities.
 */

import { z } from 'zod';

export const ProjectVersionSchema = z.object({
  id: z.string(),
  name: z.string(),
  description: z.string().nullish(),
  archived: z.boolean().default(false),
  released: z.boolean().default(false),
  startDate: z.string().nullish(),
  releaseDate: z.string().nullish(),
  overdue: z.boolean().nullish(),
  userStartDate: z.string().nullish(),
  userReleaseDate: z.string().nullish(),
  projectId: z.number(),
});

/** @typedef {z.infer<typeof ProjectVersionSchema>} ProjectVersion */

export const ProjectVersionCreateSchema = z.object({
  name: z.string(),
  description: z.string().optional(),
  projectId: z.number(),
  archived: z.boolean().default(false),
  released: z.boolean().default(false),
  startDate: z.string().optional(),
  releaseDate: z.string().optional(),
});

export const IssueTypeSchema = z.object({
  id: z.string(),
  name: z.string(),
  description: z.string().default(''),
  iconUrl: z.string().nullish(),
  subtask: z.boolean().default(false),
});

/** @typedef {z.infer<typeof IssueTypeSchema>} IssueType */

export const ProjectLeadSchema = z.object({
  accountId: z.string().nullish(),
  displayName: z.string().nullish(),
  active: z.boolean().nullish(),
  avatarUrls: z.record(z.string()).nullish(),
});

export const ProjectSchema = z.object({
  id: z.string(),
  key: z.string(),
  name: z.string(),
  description: z.string().nullish(),
  lead: ProjectLeadSchema.nullish(),
  projectTypeKey: z.string().nullish(),
  avatarUrls: z.record(z.string()).nullish(),
  url: z.string().nullish(),
  issueTypes: z.array(IssueTypeSchema).nullish(),
  versions: z.array(ProjectVersionSchema).nullish(),
});

/** @typedef {z.infer<typeof ProjectSchema>} Project */

export const ProjectDetailsSchema = z.object({
  project: ProjectSchema,
  versions: z.array(ProjectVersionSchema).default([]),
  issueTypes: z.array(IssueTypeSchema).default([]),
});
